### DISCUSSION TITLE
f-strings in python not detecting variable between "{ }"; Is this intentional?

### DISCUSSION THREAD

**Discussion by robertdefilippi** [arcticicestudio/nord#220] [Help & Support]
Hi,

I'm just going through this theme and I love it! It fits so well with the rest of my setup.

I'm just wondering about f-strings in Python and how they are formatted.

In the first image I have a f-string from my neovim using this theme.

![image](https://user-images.githubusercontent.com/9288605/104108135-3d2c1700-5290-11eb-9616-564a227b92de.png)

The second image is from my pycharm using this plugin (https://github.com/arcticicestudio/nord-jetbrains) Same repository owners, but different implementation ... I know. I just find reading the code below is much cleaner as I can differentiate between the "{}" and the variable with the curlies when writing f-strings.

![image](https://user-images.githubusercontent.com/9288605/104108199-b461ab00-5290-11eb-9687-87d9016ad659.png)

I was just wondering if I could (how to?) edit this plugin so f-strings are highlighted the same way they are in pycharm. Even if I'm pointed in the right direction, I'm sure I could work through the implementation.

Thanks,

**Comment by arcticicestudio** [comment]
Hi @robertdefilippi :wave:, thanks for your contribution :+1:

As you already said both ports are targeted for different applications which use totally different engines and parsers to detect code tokens. In case of Pycharm there would be only a single way to change it since they provide a single theme API, but for Vim there are many ways. First off there are many plugins that provide highlighting support for a specific language, but use different regular expressions to detect these.

To modify the braces we need to know if either Vim's bundled syntax parser or a third-party (plugin) is currently active. The next step would be to get the name of the highlighting group that catches the brace characters.
I use this function for the plugin development to quickly get the names of groups:

```vim
" Shows syntax highlighting groups for the current cursor position
nmap <C-S-K> :call <SID>SynStack()<CR>
function! <SID>SynStack()
  if !exists("*synstack")
    return
  endif
  echo map(synstack(line('.'), col('.')), 'synIDattr(v:val, "name")')
endfunc
```

You can add it to your `vimrc` and hold <kdb>Ctrl</kdb> and press <kdb>K</kdb> afterwards to print the chain of highlighting groups or simply change `nmap <C-S-K>` to whatever key binding you'd like to use.

**Comment by robertdefilippi** [comment]
Hi @arcticicestudio, my apologies for not responding to this sooner.

After putting in the function to my .vimrc I get the following response:

`['vimMap']`

I also ended up adding this to my `.nvim/after/syntax/python.vim` to get the highlighting correct

```
# .nvim/after/syntax/python.vim

syn match pythonEscape +{{+ contained containedin=pythonfString,pythonfDocstring
syn match pythonEscape +}}+ contained containedin=pythonfString,pythonfDocstring

syn region pythonfString matchgroup=pythonQuotes
      \ start=+[fF]\@1<=\z(['"]\)+ end="\z1"
      \ contains=@Spell,pythonEscape,pythonInterpolation

syn region pythonfDocstring matchgroup=pythonQuotes
      \ start=+[fF]\@1<=\z('''\|"""\)+ end="\z1" keepend
      \ contains=@Spell,pythonEscape,pythonSpaceError,pythonInterpolation,pythonDoctest

syn region pythonInterpolation contained
      \ matchgroup=SpecialChar
      \ start=+{{\@!+ end=+}}\@!+ skip=+{{+ keepend
      \ contains=ALLBUT,pythonDecoratorName,pythonDecorator,pythonFunction,pythonDoctestValue,pythonDoctest

syn match pythonStringModifier /:\(.[<^=>]\)\?[-+ ]\?#\?0\?[0-9]*[_,]\?\(\.[0-9]*\)\?[bcdeEfFgGnosxX%]\?/ contained containedin=pythonInterpolation
syn match pythonStringModifier /![sra]/ contained containedin=pythonInterpolation

hi link pythonfString String
hi link pyhonfDocstring String
hi link pythonStringModifier PreProc
```

I'm not sure if that is helpful, but it fixed the syntax highlighting issue I mentioned above.

**Comment by svengreb** [comment]
Thank you for your patience! 🙏🏼
It‘s been a while [since I had free time to focus more on Nord, and my open source projects in general](https://github.com/orgs/nordtheme/discussions/183#user-content-todays-scope), and invest time in this issue due to [work-life balance](https://github.com/orgs/nordtheme/discussions/183#work-life-balance).

I recently published the first [“Northern Post — The state and roadmap of Nord“ announcement](https://github.com/orgs/nordtheme/discussions/183) which includes all details about the plans and future of the _Nord_ project, including [the goal of catching up with the backlog](https://github.com/orgs/nordtheme/discussions/183#catching-up-on-the-backlog). This issue is part of the backlog and therefore I want to triage and process it to get one step closer to a “clean state“. Read the [announcement about reaching the “clean“ contribution triage state](https://github.com/orgs/nordtheme/discussions/202) in Nord‘s discussions for more details about the goal.

Due to the scope of this issue of not being “purely technical“, or being related to the project‘s code at all, I‘ll convert it into a discussion for [Nord‘s central community hub on GitHub](https://github.com/orgs/nordtheme/discussions). This way the repository is easier to maintain by separating conversations about the actual code while also making help requests, ideas and other contributions from users more visible and supportable by the community.

---

@robertdefilippi Would be nice if you could quickly check if this discussion can be closed since this is not related to this theme‘s code but the parser implementation.
